### PR TITLE
Stream writer

### PR DESCRIPTION
**Pull request by adam-fowler**
This introduces a new way to stream data into a request. Instead of asking the user for another buffer when needed via a callback a stream writer allows you to push data to a request when you want to. 
```swift
let streamWriter = AWSStreamWriter(length: 16384, on: eventLoop)
let futureResult = s3.putObject(body: .streamWriter(streamWriter), bucket: "mybucket", key: "mykey") 
streamWriter.write(.byteBuffer(buffer1))
streamWriter.write(.byteBuffer(buffer2))
streamWriter.write(.end)
try futureResult.wait()
```

In the example above we ignore the return value of `AWSStreamWriter.write`but it does return an `EventLoopFuture` which can be used to apply back pressure.

`AWSStreamWriters` can be chained together so the results of one can be processed by the next. This is used internally to implement the S3 signed streaming. The `AWSStreamWriter` results are sent to a `S3StreamWriter` which then sends its results onto the AsyncHTTPClient StreamWriter.

Now we have two full streaming solutions implemented. I hope to rewrite `StreamReader` using the new `AWSStreamWriter` instead of having it talk directly to AsyncHTTPClient. With this we can remove `S3ChunkedStreamReader` and use the chaining of StreamWriters for S3 signed `StreamReader` streams.

### CODE REVIEW

File: Sources/SotoCore/HTTP/S3StreamWriter.swift [hunk 1]
```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Soto for AWS open source project
+//
+// Copyright (c) 2017-2020 the Soto project authors
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/S3StreamWriter.swift:5]: This probably needs to be updated
```suggestion
// Copyright (c) 2017-2021 the Soto project authors
```

File: Sources/SotoCore/HTTP/S3StreamWriter.swift [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import SotoSignerV4
+
+class S3StreamWriter: StreamWriterProtocol {
+    /// Working buffer size
+    static let bufferSize: Int = 64 * 1024
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/S3StreamWriter.swift:20]: Out of curiosity why have we picked these numbers? If it's an S3 thing we should probably link to the docs

File: Sources/SotoCore/HTTP/S3StreamWriter.swift [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        self.tailBuffer = byteBufferAllocator.buffer(capacity: Self.endOfLineLength)
+        self.tailBuffer.writeString("\r\n")
+    }
+
+    /// Update HTTP headers. Add "Content-encoding" header. The "x-amz-decoded-content-length" header is added earlier in AWSRequest when
+    /// initiating the signing process
+    /// - Parameter headers: headers to update
+    func updateHeaders(headers: HTTPHeaders) -> HTTPHeaders {
+        var headers = headers
+        headers.add(name: "Content-Encoding", value: "aws-chunked")
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/S3StreamWriter.swift:71]: nit - headers _should_ be lowercased (at least in HTTP/1 - it's a must in HTTP/2 IIRC) and are lowercased internally so I suggest we follow that. Thoughts?

```suggestion
        headers.add(name: "content-Encoding", value: "aws-chunked")
```

File: Sources/SotoCore/HTTP/S3StreamWriter.swift [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        self.signingData = seedSigningData
+        // have separate buffers so we aren't allocating 128k
+        self.byteBufferAllocator = byteBufferAllocator
+        self.headerBuffer = byteBufferAllocator.buffer(capacity: Self.maxHeaderSize)
+        self.workingBuffer = byteBufferAllocator.buffer(capacity: Self.bufferSize)
+        self.tailBuffer = byteBufferAllocator.buffer(capacity: Self.endOfLineLength)
+        self.tailBuffer.writeString("\r\n")
+    }
+
+    /// Update HTTP headers. Add "Content-encoding" header. The "x-amz-decoded-content-length" header is added earlier in AWSRequest when
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/S3StreamWriter.swift:66]: See below

```suggestion
    /// Update HTTP headers. Add "content-encoding" header. The "x-amz-decoded-content-length" header is added earlier in AWSRequest when
```

File: Sources/SotoCore/HTTP/StreamWriter.swift [hunk 5]
```diff
@@ -0,0 +1,126 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Soto for AWS open source project
+//
+// Copyright (c) 2017-2020 the Soto project authors
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/StreamWriter.swift:5]: Nit

```suggestion
// Copyright (c) 2017-2021 the Soto project authors
```

File: Sources/SotoCore/HTTP/StreamWriter.swift [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    /// Write to stream writer
+    /// - Parameter result: ByteBuffer or `.end`
+    /// - Returns: Returns when EventLoopFuture for when value has been written
+    @discardableResult public func write(_ result: StreamWriterResult) -> EventLoopFuture<Void> {
+        let futureResult = self.write(result, on: self.eventLoop)
+        futureResult.cascadeFailure(to: self.finishedPromise)
+        return futureResult
+    }
+
+    /// Update headers. Add "Transfer-encoding" header if we don't have a steam size
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/StreamWriter.swift:106]: As above
```suggestion
    /// Update headers. Add "transfer-encoding" header if we don't have a steam size
```

File: Sources/SotoCore/HTTP/StreamWriter.swift [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    /// - Parameter headers: headers to update
+    func updateHeaders(headers: HTTPHeaders) -> HTTPHeaders {
+        var headers = headers
+        // add "Transfer-Encoding" header if streaming with unknown size
+        if self.length == nil {
+            headers.add(name: "Transfer-Encoding", value: "chunked")
```
Comment (0xTim) [inline Sources/SotoCore/HTTP/StreamWriter.swift:112]: As above 
```suggestion
            headers.add(name: "transfer-encoding", value: "chunked")
```

**Review by 0xTim — APPROVED** [review]
LGTM a few queries/minor nits

File: Sources/SotoCore/HTTP/S3StreamWriter.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/SotoCore/HTTP/S3StreamWriter.swift:20]: This is the recommended size from AWS. See https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

File: Sources/SotoCore/HTTP/S3StreamWriter.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/SotoCore/HTTP/S3StreamWriter.swift:71]: If we're gonna lowercase the "c" we should probably also do the same to the "e" 😝

File: Sources/SotoCore/HTTP/S3StreamWriter.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/SotoCore/HTTP/S3StreamWriter.swift]: done

File: Sources/SotoCore/HTTP/S3StreamWriter.swift (hunk 2 above)
Comment (0xTim) [reply Sources/SotoCore/HTTP/S3StreamWriter.swift:20]: I think we should add that as a comment

**Pull request closed without merging**